net/multi_listen.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package net
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+)
+
+// connErrPair pairs conn and error returned by accept on actual listeners.
+// It is used for communication between main and listener goroutines.
+type connErrPair struct {
+	conn net.Conn
+	err  error
+}
+
+// multiListener implements net.Listener
+type multiListener struct {
+	addrs     []net.Addr
+	listeners []net.Listener
+
+	wg        sync.WaitGroup
+	mu        sync.Mutex
+	connErrCh chan connErrPair
+	closed    bool
+}
+
+// compile time check to ensure *multiListener implements net.Listener.
+var _ net.Listener = &multiListener{}
+
+// MultiListen returns net.Listener which can listen for and accept
+// TCP connections on multiple addresses.
+func MultiListen(ctx context.Context, addresses []string) (net.Listener, error) {
+	return multiListen(
+		ctx,
+		addresses,
+		net.ResolveTCPAddr,
+		func(ctx context.Context, network, address string) (net.Listener, error) {
+			var lc net.ListenConfig
+			return lc.Listen(ctx, network, address)
+		})
+}
+
+// multiListen consumes stdlib dependencies as argument allowing mocking
+// for unit-tests.
+func multiListen(
+	ctx context.Context,
+	addresses []string,
+	resolveTCPAddrFunc func(network, address string) (*net.TCPAddr, error),
+	listenFunc func(ctx context.Context, network, address string) (net.Listener, error),
+) (net.Listener, error) {
+	ml := &multiListener{
+		connErrCh: make(chan connErrPair),
+	}
+
+	if len(addresses) == 0 {
+		return nil, fmt.Errorf("no address provided to listen on")
+	}
+	for _, address := range addresses {
+		addr, err := resolveTCPAddrFunc("tcp", address)
+		if err != nil {
+			// close all listeners
+			_ = ml.Close()
+			return nil, err
+		}
+
+		var network string
+		host, _, err := net.SplitHostPort(addr.String())
```

Comment:
while obviously there's something to be said for having `MultiListen` copy `net.Listen`'s API, the flip side of that is that probably most callers start out with separate IP and port, and I'm pretty sure we _never_ want the DNS-resolving behavior, and also, for our current use cases, it's always multiple IPs listening on the same port, so if `MultiListen` took an array of IPs and a single port, you'd have to do less splitting and joining later...

(Curious what other people think about "copy `net.Listen`" vs "make it better for _our_ use case".)

Reply:
My initial instinct was that it should copy `net.Listen()` but I really am OK if we decide otherwise - I just want it to be well thought out.

Passing the network argument, like `net.Listen()`, is explicit and familiar, which has some advantages, even if some of the behavior is surprising to newcomers.

Doing it "the right way" automatically is different, but that can be documented away.  Some of the magic might be surprising to expert users.

Last thing to consider: if we want to use this in k8s, we have to retain the historical semantics, which (I think?) are the Go default semantics for "tcp".  Making this library TOO smart could break kube users.

Reply:
> while obviously there's something to be said for having MultiListen copy net.Listen's API, the flip side of that is that probably most callers start out with separate IP and port, and I'm pretty sure we never want the DNS-resolving behavior, and also, for our current use cases, it's always multiple IPs listening on the same port, so if MultiListen took an array of IPs and a single port, you'd have to do less splitting and joining later...

If we remove the dns-resolving behaviour (which anyway **net.Listen** will do internally) and consume the network as an argument then there won't be any splitting and joining. 

I can't think of any use case where any consumer would want to use multi-listener for listening on different ports, and the primary use case of being able to listen on dual-stack IP pairs, I think using a list of IPs and a port will better interface. 

By using the network as an argument we don't need to handle many corner cases and the semantics will follow go defaults.  The following signature makes morse sense to me now:

```go
func MultiListen(ctx context.Context, network string, hosts []string, port int) (net.Listener, error)
```


Maybe we should also get rid of context here? as context is not fully utilized by stdlib. Cancellation/Timeout of context doesn't unblock Accept() calls. 
(ref: https://github.com/golang/go/issues/50861)

Reply:
> I can't think of any use case where any consumer would want to use multi listener for listening on different ports

Whether we can think of them or not, should we go out of our way to prevent them?  What I mean is that net.Listen() takes a host:port string, so why would we make this different on purpose?  And how will we expose this in k8s?  I presume the existing CLIs which take "host:port" would now take a list of host:port, no?  In that case the ports CAN be different, even if we don't know why someone would do that.

Reply:
FWIW, the context is verly limited utility.  It bounds the `resolve` step and is passed back to `ListenConfig.Control()` if that is defined.  It does not cancel a `Listen()` or an `Accept()`.